kmath-tensors/src/commonMain/kotlin/space/kscience/kmath/tensors/core/BufferedTensor.kt

```diff
@@ -0,0 +1,117 @@
+package space.kscience.kmath.tensors.core
+
+import space.kscience.kmath.nd.MutableBufferND
+import space.kscience.kmath.structures.*
+import space.kscience.kmath.tensors.api.TensorStructure
+import space.kscience.kmath.tensors.core.algebras.TensorLinearStructure
+
+
+public open class BufferedTensor<T>(
+    override val shape: IntArray,
+    internal val mutableBuffer: MutableBuffer<T>,
+    internal val bufferStart: Int
+) : TensorStructure<T> {
+    public val linearStructure: TensorLinearStructure
+        get() = TensorLinearStructure(shape)
+
+    public val numElements: Int
+        get() = linearStructure.linearSize
+
+    override fun get(index: IntArray): T = mutableBuffer[bufferStart + linearStructure.offset(index)]
+
+    override fun set(index: IntArray, value: T) {
+        mutableBuffer[bufferStart + linearStructure.offset(index)] = value
+    }
+
+    override fun elements(): Sequence<Pair<IntArray, T>> = linearStructure.indices().map {
+        it to this[it]
+    }
+
+    override fun equals(other: Any?): Boolean = false
```

Comment:
Isn't it broken behavior? Maybe it should be `this === other`?

Reply:
No mutable collections cannot have any notion of equality, they are all different.

Reply:
Ah OK, you mean reference equality. I don't know to be honest, I think if the user wants that she can actually use `===`, equals is kind of misleading there.